File: CS/GridViewFilter.cs
// Developer Express Code Central Example:
// How to filter the GridView across all columns and highlight the matched text
// 
// This example demonstrates how to show only those rows in the GridView whose
// cells contain a specified text. Matched text is highlighted using the
// MultiColorDrawString method.
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=E2508

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.Data.Filtering;
using DevExpress.XtraGrid.Columns;
using DevExpress.Utils.Paint;
using DevExpress.Utils;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.LookAndFeel;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;

namespace WindowsApplication1
{
    public class GridViewFilterHelper
    {
        private string _ActiveFilter = string.Empty;
        private GridView _View;
        XPaint paint = new XPaint();
       
        public string ActiveFilter
        {
            get { return _ActiveFilter; }
            set
            {
                if (_ActiveFilter != value)
                    _ActiveFilter = value;
                OnActiveFilterChanged();
            }
        }

        public GridViewFilterHelper(GridView view)
        {
            _View = view;
            _View.CustomDrawCell += new RowCellCustomDrawEventHandler(_View_CustomDrawCell);
        }

        void _View_CustomDrawCell(object sender, RowCellCustomDrawEventArgs e)
        {
            if (_ActiveFilter == string.Empty)
                return;
            int index = e.DisplayText.IndexOf(_ActiveFilter);
            if (index < 0)
                return;
            e.Handled = true;
            var inf = (e.Cell as GridCellInfo);
            TextEditViewInfo vi = inf.ViewInfo as TextEditViewInfo;
            e.Appearance.FillRectangle(e.Cache, e.Bounds);
            MultiColorDrawStringParams args = new MultiColorDrawStringParams(e.Appearance);
            args.Bounds = e.Bounds;
            args.Text = e.DisplayText;
            args.Appearance.Assign(e.Appearance);
            AppearanceDefault apperance = LookAndFeelHelper.GetHighlightSearchAppearance(vi.LookAndFeel, !vi.UseHighlightSearchAppearance);
            e.Cache.Paint.DrawMultiColorString(e.Cache, e.Bounds, e.DisplayText, _ActiveFilter, vi.PaintAppearance, vi.PaintAppearance.GetTextOptions().GetStringFormat(vi.DefaultTextOptions),
                apperance.ForeColor, apperance.BackColor, false, index);
        }

        CriteriaOperator CreateFilterCriteria()
        {
            CriteriaOperator[] operators = new CriteriaOperator[_View.VisibleColumns.Count];
            for (int i = 0; i < _View.VisibleColumns.Count; i++)
            {
                operators[i] = new BinaryOperator(_View.VisibleColumns[i].FieldName, String.Format("%{0}%", _ActiveFilter), BinaryOperatorType.Like);
            }
            return new GroupOperator(GroupOperatorType.Or, operators);

        }

        private void OnActiveFilterChanged()
        {
            _View.ActiveFilterCriteria = CreateFilterCriteria();
        }
    }
}

File: VB/GridViewFilter.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.Data.Filtering
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.Utils.Paint
Imports DevExpress.Utils
Imports DevExpress.XtraGrid.Views.Base

Namespace WindowsApplication1
	Public Class GridViewFilterHelper
		Private _ActiveFilter As String = String.Empty
		Private _View As GridView
		Private paint As New XPaint()

		Public Property ActiveFilter() As String
			Get
				Return _ActiveFilter
			End Get
			Set(ByVal value As String)
				If _ActiveFilter <> value Then
					_ActiveFilter = value
				End If
				OnActiveFilterChanged()
			End Set
		End Property

		Public Sub New(ByVal view As GridView)
			_View = view
			AddHandler _View.CustomDrawCell, AddressOf _View_CustomDrawCell
		End Sub

		Private Sub _View_CustomDrawCell(ByVal sender As Object, ByVal e As RowCellCustomDrawEventArgs)
			If _ActiveFilter = String.Empty Then
				Return
			End If
			Dim index As Integer = e.DisplayText.IndexOf(_ActiveFilter)
			If index < 0 Then
				Return
			End If
			e.Handled = True
			e.Appearance.FillRectangle(e.Cache, e.Bounds)
			Dim args As New MultiColorDrawStringParams(e.Appearance)
			args.Bounds = e.Bounds
			args.Text = e.DisplayText
			args.Appearance.Assign(e.Appearance)
			Dim apperance As AppearanceObject = _View.PaintAppearance.SelectedRow
			Dim defaultRange As New CharacterRangeWithFormat(0, e.DisplayText.Length, e.Appearance.ForeColor, e.Appearance.BackColor)
			Dim range As New CharacterRangeWithFormat(index, _ActiveFilter.Length, apperance.ForeColor, apperance.BackColor)
			args.Ranges = New CharacterRangeWithFormat() { defaultRange, range }
			paint.MultiColorDrawString(e.Cache, args)
		End Sub

		Private Function CreateFilterCriteria() As CriteriaOperator
			Dim operators(_View.VisibleColumns.Count - 1) As CriteriaOperator
			For i As Integer = 0 To _View.VisibleColumns.Count - 1
				operators(i) = New BinaryOperator(_View.VisibleColumns(i).FieldName, String.Format("%{0}%", _ActiveFilter), BinaryOperatorType.Like)
			Next i
			Return New GroupOperator(GroupOperatorType.Or, operators)

		End Function

		Private Sub OnActiveFilterChanged()
			_View.ActiveFilterCriteria = CreateFilterCriteria()
		End Sub
	End Class
End Namespace
